cranelift-frontend/src/variable.rs

```diff
@@ -1,9 +1,12 @@
 //! A basic `Variable` implementation.
 //!
-//! `FunctionBuilderContext`, `FunctionBuilder`, and related types have a `Variable`
-//! type parameter, to allow frontends that identify variables with
-//! their own index types to use them directly. Frontends which don't
-//! can use the `Variable` defined here.
+//! Frontends can use any indexing scheme they see fit and
+//! generate the appropriate `Variable` instances.
+//!
+//! Note: The `Variable` is used by Cranelift to index an array containing
+//! information about your mutable variables. Thus, when you create a new
+//! `Variable` you should make sure that the index is provided by a counter
+//! incremented by 1 each time you encounter a new mutable variable.
```

Comment:
Is this correct, @sunfishcode ? I thought Variables were meant to represent program variables, and the FunctionBuilder/FunctionBuilderContext would construct SSA values out of variables, so you don't really have to worry about having one Variable per mutable program source variable. Instead you can call `def_var` multiple times on the same variable.

Reply:
I think what is meant is that you can't just pick arbitrary indexes like `u32::MAX`, as that would cause a vec with size `u32::MAX` to be allocated. Instead you should keep the indexes as low as possible, preferably without gaps.

Reply:
Ah, makes sense. In this case, maybe let's remove all doubt for the reader; how about `The variable is used by Cranelift to index into densely allocated arrays [...]. Thus. make sure that Variable's indexes are sequentially allocated.` (or "contiguous")?